src/librustc_borrowck/borrowck/mir/elaborate_drops.rs

```diff
@@ -0,0 +1,1094 @@
+    }
+}
+
+impl<'tcx> MirPass<'tcx> for ElaborateDrops {
+    fn run_pass<'a>(&mut self, tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                    src: MirSource, mir: &mut Mir<'tcx>)
+    {
+        debug!("elaborate_drops({:?} @ {:?})", src, mir.span);
+        match src {
+            MirSource::Fn(..) => {},
+            _ => return
+        }
+        let id = src.item_id();
+        let param_env = ty::ParameterEnvironment::for_item(tcx, id);
+        let move_data = MoveData::gather_moves(mir, tcx);
+        let (elaborate_patch, _drop_flags) = {
+            let mir = &*mir;
+            let ((_, _, move_data), flow_inits) =
+                super::do_dataflow(tcx, mir, id, &[], (tcx, mir, move_data),
+                                   MaybeInitializedLvals::default());
+            let ((_, _, move_data), flow_uninits) =
+                super::do_dataflow(tcx, mir, id, &[], (tcx, mir, move_data),
+                                   MaybeUninitializedLvals::default());
+
+            match (tcx, mir, move_data) {
+                ref ctxt => ElaborateDropsCtxt {
+                    ctxt: ctxt,
+                    param_env: &param_env,
+                    flow_inits: flow_inits,
+                    flow_uninits: flow_uninits,
+                    drop_flags: FnvHashMap(),
+                    patch: MirPatch::new(mir),
+                }.elaborate()
+            }
+        };
+        pretty::dump_mir(tcx, "elaborate_drops", &0, src, mir, None);
+        elaborate_patch.apply(mir);
+        pretty::dump_mir(tcx, "elaborate_drops", &1, src, mir, None);
+    }
+}
+
+impl Pass for ElaborateDrops {}
+
+struct InitializationData {
+    live: Vec<usize>,
+    dead: Vec<usize>
+}
+
+impl InitializationData {
+    fn apply_location<'a,'tcx>(&mut self,
+                               tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                               mir: &Mir<'tcx>,
+                               move_data: &MoveData<'tcx>,
+                               loc: Location)
+    {
+        drop_flag_effects_for_location(tcx, mir, move_data, loc, |path, df| {
+            debug!("at location {:?}: setting {:?} to {:?}",
+                   loc, path, df);
+            match df {
+                DropFlagState::Live => {
+                    self.live.set_bit(path.idx());
+                    self.dead.clear_bit(path.idx());
+                }
+                DropFlagState::Dead => {
+                    self.dead.set_bit(path.idx());
+                    self.live.clear_bit(path.idx());
+                }
+            }
+        });
+    }
+
+    fn state(&self, path: MovePathIndex) -> (bool, bool) {
+        (self.live.get_bit(path.idx()), self.dead.get_bit(path.idx()))
+    }
+}
+
+impl fmt::Debug for InitializationData {
+    fn fmt(&self, _f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        Ok(())
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+/// What to do with the drop flags when we elaborate?
+enum ElaborateKind {
```

Comment:
maybe call this:

```rust
enum ClearFlags {
    All,
    None,
    This
}
```

or something? I found these names not especially intuitive (particularly `RestHead` vs `RestTail`)